File: app/src/main/java/com/mxt/anitrend/view/sheet/BottomSheetUsers.java
package com.mxt.anitrend.view.sheet;

import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;
import android.text.TextUtils;
import android.view.View;

import com.annimon.stream.IntPair;
import com.miguelcatalan.materialsearchview.MaterialSearchView;
import com.mxt.anitrend.R;
import com.mxt.anitrend.adapter.recycler.index.UserAdapter;
import com.mxt.anitrend.base.custom.sheet.BottomSheetBase;
import com.mxt.anitrend.base.custom.sheet.BottomSheetList;
import com.mxt.anitrend.databinding.BottomSheetListBinding;
import com.mxt.anitrend.model.entity.base.UserBase;
import com.mxt.anitrend.presenter.base.BasePresenter;
import com.mxt.anitrend.util.CompatUtil;
import com.mxt.anitrend.util.KeyUtil;
import com.mxt.anitrend.view.activity.detail.ProfileActivity;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;

public class BottomSheetUsers extends BottomSheetList<UserBase> implements MaterialSearchView.OnQueryTextListener, MaterialSearchView.SearchViewListener {

    private BottomSheetListBinding binding;

    public static BottomSheetUsers newInstance(Bundle bundle) {
        BottomSheetUsers fragment = new BottomSheetUsers();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = new BasePresenter(getContext());
        mColumnSize = getResources().getInteger(R.integer.single_list_x1);
        mAdapter = new UserAdapter(getActivity());
        if(getArguments() != null) {
            List<UserBase> baseList = getArguments().getParcelableArrayList(KeyUtil.arg_list_model);
            if(!CompatUtil.INSTANCE.isEmpty(baseList))
                mAdapter.onItemsInserted(baseList);
        }
    }

    /**
     * Setup your view un-binder here as well as inflating other views as needed
     * into your view stub
     *
     * @param savedInstanceState
     */
    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        binding = BottomSheetListBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(getActivity()));
        dialog.setContentView(binding.getRoot());
        unbinder = ButterKnife.bind(this, dialog);
        createBottomSheetBehavior(binding.getRoot());
        mLayoutManager = new StaggeredGridLayoutManager(mColumnSize, StaggeredGridLayoutManager.VERTICAL);
        return dialog;
    }

    /**
     * Is automatically called in the @onStart Method if overridden in list implementation
     */
    @Override
    protected void updateUI() {
        toolbarTitle.setText(getString(mTitle, mAdapter.getItemCount()));
        toolbarSearch.setVisibility(View.VISIBLE);
        searchView.setOnQueryTextListener(this);
        searchView.setOnSearchViewListener(this);
        injectAdapter();
    }

    /**
     * All new or updated network requests should be handled in this method
     */
    @Override
    public void makeRequest() {

    }

    /**
     * Called when the user submits the query. This could be due to a key press on the
     * keyboard or due to pressing a submit button.
     * The listener can override the standard behavior by returning true
     * to indicate that it has handled the submit request. Otherwise return false to
     * let the SearchView handle the submission by launching any associated intent.
     *
     * @param query the query text that is to be submitted
     * @return true if the query has been handled by the listener, false to let the
     * SearchView perform the default action.
     */
    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    /**
     * Called when the query text is changed by the user.
     *
     * @param newText the new content of the query text field.
     * @return false if the SearchView should perform the default action of showing any
     * suggestions if available, true if the action was handled by the listener.
     */
    @Override
    public boolean onQueryTextChange(String newText) {
        if(!TextUtils.isEmpty(newText) && mAdapter != null && mAdapter.getFilter() != null) {
            mAdapter.getFilter().filter(newText);
            return true;
        }
        return false;
    }

    @Override
    public void onSearchViewShown() {
        bottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
    }

    @Override
    public void onSearchViewClosed() {
        if(mAdapter != null && mAdapter.getFilter() != null)
            mAdapter.getFilter().filter("");
    }

    /**
     * When the target view from {@link View.OnClickListener}
     * is clicked from a view holder this method will be called
     *
     * @param target view that has been clicked
     * @param data   the model that at the clicked index
     */
    @Override
    public void onItemClick(View target, IntPair<UserBase> data) {
        switch (target.getId()) {
            case R.id.container:
                Intent intent = new Intent(getActivity(), ProfileActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                intent.putExtra(KeyUtil.arg_id, data.getSecond().getId());
                if(getActivity() != null)
                    getActivity().startActivity(intent);
                break;
        }
    }

    /**
     * When the target view from {@link View.OnLongClickListener}
     * is clicked from a view holder this method will be called
     *
     * @param target view that has been long clicked
     * @param data   the model that at the long clicked index
     */
    @Override
    public void onItemLongClick(View target, IntPair<UserBase> data) {

    }

    /**
     * Builder class for bottom sheet
     */
    public static class Builder extends BottomSheetBuilder {

        @Override
        public BottomSheetBase build() {
            return newInstance(bundle);
        }

        public BottomSheetBuilder setModel(List<UserBase> model) {
            bundle.putParcelableArrayList(KeyUtil.arg_list_model, (ArrayList<? extends Parcelable>) model);
            return this;
        }
    }
}

File: app/src/main/java/com/mxt/anitrend/view/sheet/BottomSheetUsers.kt
package com.mxt.anitrend.view.sheet

import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.mxt.anitrend.R
import com.mxt.anitrend.adapter.recycler.index.UserAdapter
import com.mxt.anitrend.base.custom.recycler.RecyclerViewAdapter
import com.mxt.anitrend.base.custom.recycler.StatefulRecyclerView
import com.mxt.anitrend.base.custom.sheet.BottomSheetBase
import com.mxt.anitrend.base.interfaces.event.ISearchDelegate
import com.mxt.anitrend.base.interfaces.event.ItemClickListener
import com.mxt.anitrend.data.DatabaseHelper
import com.mxt.anitrend.data.store.user.UserStore
import com.mxt.anitrend.databinding.BottomSheetListBinding
import com.mxt.anitrend.domain.model.ToggleUserFollowCommand
import com.mxt.anitrend.domain.model.UserRecord
import com.mxt.anitrend.domain.user.interactor.ToggleUserFollowInteractor
import com.mxt.anitrend.extension.getCompatDrawable
import com.mxt.anitrend.extension.parcelableArrayList
import com.mxt.anitrend.model.entity.base.UserBase
import com.mxt.anitrend.util.KeyUtil
import com.mxt.anitrend.view.activity.detail.ProfileActivity
import com.mxt.anitrend.widget.ProgressLayout
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject

class BottomSheetUsers :
    BottomSheetBase<List<UserBase>>(),
    ItemClickListener<UserBase>,
    Observer<List<UserBase>?> {

    private var binding: BottomSheetListBinding? = null

    private lateinit var mAdapter: RecyclerViewAdapter<UserBase>
    private lateinit var mLayoutManager: StaggeredGridLayoutManager

    private var stateLayout: ProgressLayout? = null
    private var recyclerView: StatefulRecyclerView? = null
    private var mColumnSize: Int = 0

    private val databaseHelper: DatabaseHelper by inject()
    private val userStore: UserStore by inject()
    private val toggleUserFollowInteractor: ToggleUserFollowInteractor by inject()

    companion object {
        @JvmStatic
        fun newInstance(bundle: Bundle): BottomSheetUsers = BottomSheetUsers().apply {
            arguments = bundle
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val ctx = requireContext()
        mColumnSize = resources.getInteger(R.integer.single_list_x1)
        mAdapter = UserAdapter(
            context = ctx,
            currentUser = databaseHelper.currentUser,
            onToggleFollowAction = ::toggleFollow,
        )
        val baseList = arguments?.parcelableArrayList<UserBase>(KeyUtil.arg_list_model)
        if (!baseList.isNullOrEmpty()) {
            mAdapter.onItemsInserted(baseList)
        }
        observeUserStore()
    }

    /**
     * Fire-and-forget delivery from the render-only [FollowStateWidget]. The legacy adapter
     * still passes a result callback slot, which is intentionally ignored: the mutation result
     * is applied by observing [UserStore] below.
     */
    private fun toggleFollow(
        userId: Long,
        @Suppress("UNUSED_PARAMETER") onResult: (Result<UserBase>) -> Unit,
    ) {
        lifecycleScope.launch {
            toggleUserFollowInteractor(ToggleUserFollowCommand(userId = userId))
        }
    }

    /**
     * Smallest behavior-preserving bridge between the canonical [UserStore] and the legacy
     * adapter: committed follow changes for users in the list are written back onto the
     * existing item so the widget re-renders the authoritative state.
     */
    private fun observeUserStore() {
        lifecycleScope.launch {
            userStore.state.collect { state ->
                state.usersById.values.forEach(::rebindUserIfPresent)
            }
        }
    }

    private fun rebindUserIfPresent(record: UserRecord) {
        val position = mAdapter.data.indexOfFirst { it.id == record.id }
        if (position < 0) return
        val current = mAdapter.data[position]
        if (current.isFollowing == record.isFollowing) return
        current.isFollowing = record.isFollowing
        mAdapter.onItemChanged(current, position)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        binding = BottomSheetListBinding.inflate(layoutInflater)
        dialog.setContentView(requireNotNull(binding).root)
        bindToolbarViews(requireNotNull(binding).root)
        searchView = binding?.customSheetToolbar?.searchView
        stateLayout = binding?.stateLayout
        recyclerView = binding?.recyclerView
        createBottomSheetBehavior(requireNotNull(binding).root)
        mLayoutManager = StaggeredGridLayoutManager(mColumnSize, StaggeredGridLayoutManager.VERTICAL)
        return dialog
    }

    override fun onStart() {
        super.onStart()
        toolbarTitle?.text = getString(mTitle, mAdapter.itemCount)
        toolbarSearch?.visibility = View.VISIBLE
        mSearchDelegate = object : ISearchDelegate {
            override fun onQueryChanged(query: String?) {
                if (!TextUtils.isEmpty(query) && mAdapter.filter != null) {
                    mAdapter.filter.filter(query)
                }
            }

            override fun onSearchSubmitted(query: String?) = Unit

            override fun onSearchShown() {
                bottomSheetBehavior?.state = BottomSheetBehavior.STATE_EXPANDED
            }

            override fun onSearchClosed() {
                mAdapter.filter?.filter("")
            }
        }
        injectAdapter()
    }

    private fun injectAdapter() {
        mAdapter.setClickListener(this)
        val recycler = recyclerView ?: return
        if (recycler.adapter == null) {
            recycler.setHasFixedSize(true)
            recycler.isNestedScrollingEnabled = true
            recycler.layoutManager = mLayoutManager
            recycler.adapter = mAdapter
        }
        if (mAdapter.itemCount < 1) {
            val drawable =
                context?.getCompatDrawable(
                    R.drawable.ic_new_releases_white_24dp,
                    R.color.colorStateBlue,
                ) ?: return
            stateLayout?.showEmpty(drawable, getString(R.string.layout_empty_response))
        } else {
            stateLayout?.showContent()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onChanged(value: List<UserBase>?) = Unit

    override fun onItemClick(
        target: View,
        data: IndexedValue<UserBase>,
    ) {
        when (target.id) {
            R.id.container -> {
                val host = activity ?: return
                val intent =
                    Intent(host, ProfileActivity::class.java).apply {
                        flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        putExtra(KeyUtil.arg_id, data.value.id)
                    }
                host.startActivity(intent)
            }
        }
    }

    override fun onItemLongClick(
        target: View,
        data: IndexedValue<UserBase>,
    ) = Unit

    class Builder : BottomSheetBuilder() {
        override fun build(): BottomSheetBase<*> = newInstance(bundle)

        fun setModel(model: List<UserBase>): Builder {
            bundle.putParcelableArrayList(KeyUtil.arg_list_model, ArrayList(model))
            return this
        }
    }
}
